adapters/sharethrough/butler.go

```diff
@@ -0,0 +1,193 @@
+	Height             uint64
+	Width              uint64
+}
+
+type StrOpenRTBInterface interface {
+	requestFromOpenRTB(openrtb.Imp, *openrtb.BidRequest) (*adapters.RequestData, error)
+	responseToOpenRTB(openrtb_ext.ExtImpSharethroughResponse, *adapters.RequestData) (*adapters.BidderResponse, []error)
+}
+
+type StrAdServerUriInterface interface {
+	buildUri(StrAdSeverParams) string
+	parseUri(string) (*StrAdSeverParams, error)
+}
+
+type UserAgentParsers struct {
+	ChromeVersion    *regexp.Regexp
+	ChromeiOSVersion *regexp.Regexp
+	SafariVersion    *regexp.Regexp
+}
+
+type StrUriHelper struct {
+	BaseURI string
+}
+
+type StrOpenRTBTranslator struct {
+	UriHelper        StrAdServerUriInterface
+	Util             UtilityInterface
+	UserAgentParsers UserAgentParsers
+}
+
+func (s StrOpenRTBTranslator) requestFromOpenRTB(imp openrtb.Imp, request *openrtb.BidRequest) (*adapters.RequestData, error) {
+	headers := http.Header{}
+	headers.Add("Content-Type", "text/plain;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	var strImpExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &strImpExt); err != nil {
+		return nil, err
+	}
+	var strImpParams openrtb_ext.ExtImpSharethroughExt
+	if err := json.Unmarshal(strImpExt.Bidder, &strImpParams); err != nil {
+		return nil, err
+	}
+
+	pKey := strImpParams.Pkey
+
+	var height, width uint64
+	if len(strImpParams.IframeSize) >= 2 {
+		height, width = uint64(strImpParams.IframeSize[0]), uint64(strImpParams.IframeSize[1])
+	} else {
+		height, width = s.Util.getPlacementSize(imp.Banner.Format)
+	}
+
+	return &adapters.RequestData{
+		Method: "POST",
+		Uri: s.UriHelper.buildUri(StrAdSeverParams{
+			Pkey:               pKey,
+			BidID:              imp.ID,
+			ConsentRequired:    s.Util.gdprApplies(request),
+			ConsentString:      s.Util.gdprConsentString(request),
+			Iframe:             strImpParams.Iframe,
+			Height:             height,
+			Width:              width,
+			InstantPlayCapable: s.Util.canAutoPlayVideo(request.Device.UA, s.UserAgentParsers),
+		}),
+		Body:    nil,
+		Headers: headers,
+	}, nil
+}
+
+func (s StrOpenRTBTranslator) responseToOpenRTB(strResp openrtb_ext.ExtImpSharethroughResponse, btlrReq *adapters.RequestData) (*adapters.BidderResponse, []error) {
+	var errs []error
+	bidResponse := adapters.NewBidderResponse()
+
+	bidResponse.Currency = "USD"
+	typedBid := &adapters.TypedBid{BidType: openrtb_ext.BidTypeNative}
+
+	if len(strResp.Creatives) == 0 {
+		errs = append(errs, &errortypes.BadInput{Message: "No creative provided"})
+		return nil, errs
+	}
+	creative := strResp.Creatives[0]
+
+	btlrParams, parseHBUriErr := s.UriHelper.parseUri(btlrReq.Uri)
+	if parseHBUriErr != nil {
+		errs = append(errs, &errortypes.BadInput{Message: parseHBUriErr.Error()})
+		return nil, errs
+	}
+
+	adm, admErr := s.Util.getAdMarkup(strResp, btlrParams)
+	if admErr != nil {
+		errs = append(errs, &errortypes.BadServerResponse{Message: admErr.Error()})
+	}
+	if adm == "" {
```

Comment:
From the code in `getAdMarkup`, it looks like, `adm` would be empty only if there was an error. So I'd suggest `return nil, errs` be moved to the if block above because now you have two checks, one, if `getAdMarkup` returns an error and another if `adm` is empty when they are likely gonna be true together. If there's a chance that they might not be true together, I'd change the code to be:
```
if admErr != nil {
	errs = append(errs, &errortypes.BadServerResponse{Message: admErr.Error()})
	return nil, errs
}
if adm == "" {
	errs = append(errs, &errortypes.BadServerResponse{Message: "ADM can't be empty"})
	return nil, errs
}
```

Reply:
Fair point, I merged them in a single test and return. Thanks